core/src/test/java/org/elasticsearch/common/cache/CacheTests.java

```diff
@@ -0,0 +1,473 @@
+                return key;
+            }
+        }
+        Cache<Key, String> cache = CacheBuilder.<Key, String>builder().weigher((k, s) -> k.weight).build();
+        for (int i = 0; i < numberOfEntries; i++) {
+            cache.put(new Key(i, 1), Integer.toString(i));
+        }
+        assertEquals(numberOfEntries, cache.count());
+        assertEquals(numberOfEntries, cache.weight());
+        int replaced = 0;
+        for (int i = 0; i < numberOfEntries; i++) {
+            if (rarely()) {
+                replaced++;
+                cache.put(new Key(i, 2), Integer.toString(i));
+            }
+        }
+        assertEquals(numberOfEntries, cache.count());
+        assertEquals(numberOfEntries + replaced, cache.weight());
+    }
+
+    // randomly replace some entries, then check that we received replacement notifications for those and only those
+    // entries
+    public void testNotificationOnReplace() {
+        Set<Integer> notifications = new HashSet<>();
+        Cache<Integer, String> cache =
+                CacheBuilder.<Integer, String>builder()
+                        .removalListener(notification -> {
+                            assertEquals(RemovalNotification.RemovalReason.REPLACED, notification.getRemovalReason());
+                            notifications.add(notification.getKey());
+                        })
+                        .build();
+        for (int i = 0; i < numberOfEntries; i++) {
+            cache.put(i, Integer.toString(i));
+        }
+        Set<Integer> replacements = new HashSet<>();
+        for (int i = 0; i < numberOfEntries; i++) {
+            if (rarely()) {
+                cache.put(i, Integer.toString(i) + Integer.toString(i));
+                replacements.add(i);
+            }
+        }
+        assertEquals(replacements, notifications);
+    }
+
+    // test that the cache is not corrupted under lots of concurrent modifications, even hitting the same key
+    // here be dragons: this test did catch one subtle bug during development; do not remove lightly
+    public void testTorture() throws InterruptedException {
+        int numberOfThreads = randomIntBetween(1, 200);
+        AtomicInteger count = new AtomicInteger();
+        final Cache<Integer, String> cache =
+                CacheBuilder.<Integer, String>builder()
+                        .setMaximumWeight(1000)
+                        .removalListener(notification -> count.decrementAndGet())
+                        .weigher((k, v) -> 2)
+                        .build();
+
+        List<Thread> threads = new ArrayList<>();
+        for (int i = 0; i < numberOfThreads; i++) {
+            Thread thread = new Thread(() -> {
+                for (int j = 0; j < numberOfEntries; j++) {
+                    Integer key = randomIntBetween(1, numberOfEntries);
+                    cache.put(key, randomAsciiOfLength(10));
+                    count.incrementAndGet();
+                    if (rarely()) {
```

Comment:
while i know some of these helper methods are friendly, when testing concurrency for something like this, that will be used everywhere, I would remove any other concurrency from the equation: otherwise you have happens-befores that could hide bugs.

All of the lines above cause synchronization: calls to `random()`, either explicitly or implicitly via `randomIntBetween`, `randomAsciiOfLength`, `rarely` etc synchronize on a global lock in `RandomizedContext`.
The `AtomicInteger` seems unnecessary, can we just change this assert at the end to compare against `numberOfEntries * numberOfThreads` ?